Add support to CF-TLS for scanning elliptic curve support in CFSSL.

jkroll opened the pull request:
Specifically:
-Add the IANA-named curves (and the curves for the CHACHA20_POLY1305 suites, which aren't IANA managed but are effectively standards).
-Add support to the cfsslscan helper functions to move the handshake beyond ServerHello. Parse our the server's preferred curve ID and pass it back to CF-SSL for use in the scan.

jkroll committed:
Add support for scanning elliptic curve support in CFSSL.

Specifically:
-Add the IANA-named curves (and the curves for the CHACHA20_POLY1305 suites, which aren't IANA managed but are effectively standards).
-Add support to the cfsslscan helper functions to move the handshake beyond ServerHello. Parse our the server's preferred curve ID and pass it back to CF-SSL for use in the scan.

0xhaven commented on tls/cfsslscan_handshake.go:
```diff
@@ -2,9 +2,58 @@ package tls
+// EC-based TLS cipher suite. Comments indicate non-EC ranges.
+// encodings are defined by IANA:
+// http://www.iana.org/assignments/tls-parameters/tls-parameters.xml
+func IsECcipherSuite(cs uint16) bool {
```
I think I'd prefer there in the `CipherSuites` lookup table rather than as a function. i.e. add an `CipherSuite.EllipticCurve` field like the `CipherSuite.ForwardSecret` field, or the `suiteECDHE` flag in the [crypto/tls](https://github.com/cloudflare/cf-tls/blob/master/tls/cipher_suites.go#L38) cipher suites.

In fact, `crypto/tls` is probably best with just a generic flags field like that has possible values from constants `EllipticCurve`, `ForwardSecret`, etc.

jkroll replied:
This logic does look bizzare, and I agree that a table is cleaner. I did it this way because, originally (and as a hack) we were simply masking the cipherID against 0xc000, since that works for all the Go-TLS-supported suites and I was trying to extend that logic to cover all the cases. But you're right that using a table is the best solution, and I've ported the information over to CipherSuite and nuked this function.

0xhaven commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +71,51 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
+		var skx *serverKeyExchangeMsg
+ 		skx, err = c.exchangeKeys()
+		if (skx.raw[0] != typeServerKeyExchange) {
+			panic("Bad handshake! Did not get ServerKeyExchange when expected!")
```
![Don't Panic](https://upload.wikimedia.org/wikipedia/commons/5/5b/Towel_dna.jpg)
Just return an error message. 

```
err = errors.New("tls: didn't receive expected ServerKeyExchange message")
return
```

jkroll replied:
I was trying to make extra-double sure that I was parsing the right message (while trying to debug the unstable scan results) and this method seemed more readable than the awful reflect-based things you have to do to get a type assertion about the value of a pointer. But I've replaced it with an error and convinced myself that (a) I really am parsing ServerKeyExchange (b) I'm parsing it correctly (or at least consistently with processServerKeyExchange in key_agreement.go) and (c) this check is basically redundant given the type assertion and the logic that checks the message's magic number in readHandshake (see conn.go).

For now, I've replaced the panic with an error, but it probably makes sense to remove the check altogether, in fact.

0xhaven commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +71,51 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
+		if (skx.raw[0] != typeServerKeyExchange) {
+			panic("Bad handshake! Did not get ServerKeyExchange when expected!")
+		}
+ 		if err != nil {
```
Check error conditions before operating on data. `err != nil` could imply `skx.raw` is empty.

jkroll replied:
Added the length check from processServerKeyExchange in key_agreement.go.

0xhaven replied:
My point was `err != nil` probably implies `skx == nil` or `skx.raw` is empty.

This came up when testing against an unresponsive serve (`err` due to timeout and `skx==nil`) and caused scan to crash :)

jkroll replied:
The check in processServerKeyExchange explicitly raises an error on short, non-nil skx messages:

```
if len(skx.key) < 4 {
       err = errServerKeyExchange
      return
 }
```

My guess is that the 4 is there because RFC4346 specifies the first four bytes as a message header. They're thrown out during unmarshaling of ServerKeyExchange messages - that's the only difference between `skx.raw` and `skx.key`.

I believe the `skx==nil` case is checked during unmarshaling.

0xhaven commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +71,51 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
+			err = unexpectedMessageError(certMsg, msg)
+			return
+		}
+		if serverHello.ocspStapling {
```
We don't really care is the serve is trying to use the OCSP Stapling (has set the flag in their hello), so we can just remove all of this logic.

jkroll replied:
So it doesn't work if you don't read the Certificate message out of the handshake prior to reading ServerKeyExchange, and I feel like the error checks are useful. Given that, I also feel like if there's a stapled CertificateStatus message, it will also need to be read, even if we don't care about it.

I suppose we could simplify the logic in here by removing the type assertions and the checks about them, since they're redundant with logic in readHandshake. But readHandshake still must be called here, and an extra time if the server is doing OCSP Stapling.

0xhaven replied:
Ah, that's reasonable that Go need to read the messages off the stream if they exist, should be fine.

0xhaven commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
+		
+		var skx *serverKeyExchangeMsg
+ 		skx, err = c.exchangeKeys()
+		if (skx.raw[0] != typeServerKeyExchange) {
```
Style nit: in Go you don't use parens around control statement predicates (unless necessary).
I'd grep for all the `if (` instances and remove them.

Other than that, everything looks good!

kisom replied:
`gofmt` (or better yet, `goimports`) should catch that, too&mdash;it's a good idea to run it over your code before committing.

jkroll replied:
Have run code through `go fmt` (both here and in the pull request for cfssl). It does indeed remove brackets around if conditions. Also, apparently I'm kind of a space cadet this week for forgetting that.

jkroll commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
-
 }
 
 // sayHello is the backend to SayHello that returns a full serverHelloMsg for processing.
```
Sometimes (commonly, but not uniformly), this function (specifically, the call to `c.readHandshake()`) returns the error `remote error: handshake failure` which is passed all the way out to the output. I can't discern any pattern in the requests that cause this error, other than that all the requests that trigger it have (a) only an EC cipher suite requested and (b) all possible curves. But not all such requests trigger the error.

grittygrease replied:
Which sites cause this error? It might be caused by a bug in some F5 load balancers (https://www.imperialviolet.org/2014/02/27/tlssymmetriccrypto.html).

> But it was the F5 workaround that caused most of the headaches. Older (but common) firmware revisions of F5 devices hang the connection when the ClientHello is longer than 255 bytes. This appeared to be a major issue because it meant that we couldn't add things to the ClientHello without breaking lots of big sites that use F5 devices. After a plea and a discussion on the TLS list, an engineer from F5 explained that it was just handshake messages between 256 and 511 bytes in length that caused the issue. That suggested an obvious solution: pad the handshake message so that it doesn't fall into the troublesome size range.

jkroll replied:
I see it on e.g. cloudflare.com. I also guessed it had something to do with large clientHello messages, but I don't think that's correct - it doesn't happen for every request. And the connections don't hang - they send a TLS "Alert" message. So perhaps it's really happening when an ECDH cipher suite is requested but no requested curves are available.

From RFC 5246:

> handshake_failure
>      Reception of a handshake_failure alert message indicates that the
>      sender was unable to negotiate an acceptable set of security
>      parameters given the options available.  This is a fatal error.

I've went throught and characterized the errors pretty well: there's exactly one of these messages for each `(serverCipher, TLSVersion)` pair negotiated by the server (where `tls.CipherSuites[serverCipher].EllipticCurve == true`), which suggests to me that this is happening in response to the server actually not being able to negotiate anything (which is odd, because the `clientHello` corresponding to the `Alert` message actually contains a negotiable set of parameters, namely all the named curves, at least one of which should be valid). This led me to wonder if you're not allowed to request the curveID values that correspond to non-named curves, and you aren't, per http://tools.ietf.org/html/rfc4492#section-5.4. Dropping those values from the initial set of `tls.CurveID`s presented didn't change where the errors occur, though.

There is also 1 additional error per TLSVersion, which has _most_ of the CipherSuites (315 out of 320) and all of the curve options.

grittygrease commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
+		}
+
+		var skx *serverKeyExchangeMsg
+		skx, err = c.exchangeKeys()
```
The error should be checked before dereferencing skx.

grittygrease commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
+			err = errors.New("tls: received something other than the expected ServerKeyExchange.")
+			return
+		}
+		if err != nil {
```
Move this check up to before line 55.

jkroll replied:
Good point. Somehow I thought more checking was done during message unmarshaling, but checking there is minimal - most errors are passed through and the library seems to defer checking to the caller (e.g., this check is not present in `processServerKeyExchange` in key_agreement.go, but is performed by the caller in `doFullHandshake` in handshake_client.go).

0xhaven replied:
This is pretty standard in Go, minimal checking followed by short circuiting that passes errors up as high on the stack as possible before being handled with some action (logging, sending messages, stopping the goroutine, etc.)

jkroll commented:
Sorry - the print statements were something I added to characterize the alert messages. I forgot to remove them before pushing. I've removed them now.

0xhaven commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
+		}
+		certMsg, ok := msg.(*certificateMsg)
+		if !ok || len(certMsg.certificates) == 0 {
+			c.sendAlert(alertUnexpectedMessage)
```
Similarly, this `sendAlert` is unnecessary.

0xhaven commented on tls/cfsslscan_handshake.go:
```diff
@@ -22,9 +22,55 @@ func (c *Conn) SayHello() (cipherID, version uint16, err error) {
+			if err != nil {
+				return
+			}
+			_, ok := msg.(*certificateStatusMsg)
```
These should all just follow the same (slight variation) of the same format:

```
    if someMsg, ok := msg.(*expectedType); !ok {
        // REMOVED: c.sendAlert(alertUnexpectedMessage)
        err = unexpectedMessageError(someMsg, msg)
        return
    }
```

grittygrease commented:
This looks good modulo Jacob's comments. Please address the final ones and we will merge it in.

jkroll commented:
I have removed the sendAlert calls, per Jacob's suggestion, and confirmed using grep that there are no such calls anywhere in the cfssl support code. Should still catch the expected alert messages in the curve scanning code, which I'll take a look at this afternoon.

0xhaven commented on tls/conn.go:
```diff
@@ -626,7 +628,13 @@ Again:
 		case alertLevelError:
-			c.in.setErrorLocked(&net.OpError{Op: "remote error", Err: alert(data[1])})
+			var e error
+			if alert(data[1]) == alertHandshakeFailure {
```
This if statement doesn't actually change the behavior of the program, so I don't think it needs to be added here. I would very much prefer not to modify original `crypto/tls` library code as it will make merging changes from upstream more complicated.

jkroll replied:
This seemed simpler and more like a common Go pattern than some sort of deep comparison, but see the code depending on it in cfssl/scan/tls_handshake.go.

0xhaven commented on tls/conn.go:
```diff
@@ -19,6 +19,8 @@ import (
 	"time"
 )
 
+var HandshakeFailureErr = &net.OpError{Op: "remote error", Err: alert(alertHandshakeFailure)}
```
If we want to export the form of the error (mainly the problem just being the opaque `alertHandshakeFailure`), we can do so in one of the helper files. The actual comparison may or may not need to be done using [reflect.DeepEqual](https://golang.org/pkg/reflect/#DeepEqual).

jkroll replied:
Understood. The control flow in the scan is odd: the scan loop terminates based on the receipt of an alert message, which causes the scan to always print an error. I want to catch this error and ignore it. The only reliable way to do this is to change the way alert messages are handled.

That solution isn't totally ideal, because it means that I'm overloading alertHandshakeFailure to cover (1) the expected failure at the end of the scan, when no viable parameters can be chosen and (2) any other event which would cause that alert message. In particular, once I catch that error in the scan in cfssl, I simply assign `err = nil` and move on, which effectively papers over some legitimate error cases. I've pushed that code, so you can see what I was trying to do. I'd love to talk about the right way to do this, though.

reflect.DeepEqual solves the particular comparison problem, but not the overloading problem. But it may be the best option.

jkroll commented:
So  I've reverted the commits since 877a71762490c938f39f5855af4a5e42a38e3c4c. Is that the way you want the history structured or should I cherry-pick the good stuff onto a new branch with a cleaner history?

0xhaven commented:
The simplest thing would probably be just a rebase down to a single, descriptive commit message `git rebase -i 3c5ad137e89d9a6344b1f2b6b9d079a551376c8e`, choosing "fixup" to discard all of the commit messages that are just cleaning things up. (Or you can edit anything new you think is important into the primary commit message):

```
pick e8ee3c4 Add support for scanning elliptic curve support in CFSSL.
f 9a11d77 Replace the wacky logic for determining elliptic-curve-ness of a cipher suite with more sensible entries in the CipherSuite table.
f c6e21f9 Replace an ill-advised panic with an error message.
f 50a13aa Check error conditions consistently with processServerKeyExchange in key_agreement.go
f 6faec31 Compute the curve id consistently with processServerKeyExchange in key_agreement.go.
f fcddfdd gofmt.
f 2d23836 Move an error check so that it happens prior to any use of the related value.
f a6722d5 Remove unnecessary print statements.
f 7f52f5f Remove unnecessary alert messages to speed scanning.
f 877a717 Return all errors in the same way when messages don't parse correctly.

# [All of these are discarded...]
# 1355319 Modify the handling of alert messages so we can test for alertHandshakeFailure during scans, as it's the expected termination condition for the scanning loop.
# 6c2defe Revert "Modify the handling of alert messages so we can test for alertHandshakeFailure during scans, as it's the expected termination condition for the scanning loop."
# ff53c9e Add a sentinel to help detect expected handshake failures during scan.
# 5f4f635 Revert "Add a sentinel to help detect expected handshake failures during scan."

```

And then `git push -f` to update the PR and it'll end up looking [something like this](https://github.com/cloudflare/cf-tls/commit/076234cbc973333abd57f196a6a29e4866ee246a)

jkroll commented:
I didn't know you could push -f to a PR and not end up in a strange state. Cool!

0xhaven commented:
Strange is relative, you do get a bunch of "X commented on an outdated diff", but that's fine.
